### ISSUE TITLE
differentiable setitem

### ISSUE THREAD

**Issue by chenyuxyz** [tinygrad/tinygrad#4574]
example usage from discord
```
for t in range(0, inp_seq_len):
    model(X[:, t].to(device))

y_pred = torch.zeros(Y.size()).to(device)
for i in range(0, out_seq_len):
    y_pred[:, i] = model()

loss = criterion(y_pred.to(device), Y.to(device))
loss.backward()
clip_grads(model)
optimizer.step()
```

setitem currently works if the set area can be represented with a shapetracker, and it partial assign using the said shapetracker as output shapetracker.

to make it differentiable, we can convert that shapetracker into a mask tensor, then use where to construct the new tensor

**Comment by chenyuxyz** [comment]
tricky if first set with tensor with grad, then partially set with tensor without grad...

**Comment by geohotstan** [comment]
> to make it differentiable, we can convert that shapetracker into a mask tensor, then use where to construct the new tensor

#4489 this does it by appending an indexedview to the shapetracker that describes the indexing rules of the buffer when loading or storing. Is it kinda on the right track? Just have to implement indexing functions in function.py and maybe it works? It also works for overlapped cases.

**Comment by chenyuxyz** [comment]
i want to do the no-copy case without adding indexedview. maybe it's needed for advance setitem? not sure.

say `t[idx] = a`
`idx` became shapetracker `st` (how getitem works), then we apply inverse of `st` to `a` -> `full_a = inv(st, t.shape)(a)` and `ones_like(a)` -> `mask = inv(st, t.shape)(ones_like(a))`. and update `t = mask.where(full_a, t)` should work

**Comment by chenyuxyz** [comment]
the st inversion part and masked tensor parts are fine (`LazyBuffer._view` is great), the issue now is that grad does not flow through assign

**Comment by chenyuxyz** [comment]
torch behavior
```
a = torch.ones((8, 8))
b = torch.rand((8, 8), requires_grad=True)
a[:] = b
a.sum().backward()
print(b.grad) # correct
print(a.grad) # warning and returned None

a = torch.ones((8, 8), requires_grad=True)
b = torch.rand((8, 8))
a[:] = b # not allowed
```
